internal/controller/rabbitmq/rabbitmqpolicy_controller.go

```diff
@@ -0,0 +1,247 @@
+	rabbitmqapi "github.com/openstack-k8s-operators/infra-operator/pkg/rabbitmq/api"
+	condition "github.com/openstack-k8s-operators/lib-common/modules/common/condition"
+	helper "github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	oko_secret "github.com/openstack-k8s-operators/lib-common/modules/common/secret"
+	rabbitmqclusterv2 "github.com/rabbitmq/cluster-operator/v2/api/v1beta1"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+)
+
+const policyFinalizer = "rabbitmqpolicy.openstack.org/finalizer"
+
+// RabbitMQPolicyReconciler reconciles a RabbitMQPolicy object
+//
+//nolint:revive
+type RabbitMQPolicyReconciler struct {
+	client.Client
+	Kclient kubernetes.Interface
+	Scheme  *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=rabbitmq.openstack.org,resources=rabbitmqpolicies,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=rabbitmq.openstack.org,resources=rabbitmqpolicies/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=rabbitmq.openstack.org,resources=rabbitmqpolicies/finalizers,verbs=update
+
+// Reconcile reconciles a RabbitMQPolicy object
+func (r *RabbitMQPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	Log := log.FromContext(ctx)
+
+	instance := &rabbitmqv1.RabbitMQPolicy{}
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	h, _ := helper.NewHelper(instance, r.Client, r.Kclient, r.Scheme, Log)
+
+	// Save a copy of the conditions so that we can restore the LastTransitionTime
+	// when a condition's state doesn't change
+	savedConditions := instance.Status.Conditions.DeepCopy()
+
+	// Initialize status conditions
+	cl := condition.CreateList(
+		condition.UnknownCondition(condition.ReadyCondition, condition.InitReason, condition.ReadyInitMessage),
+		condition.UnknownCondition(rabbitmqv1.RabbitMQPolicyReadyCondition, condition.InitReason, rabbitmqv1.RabbitMQPolicyReadyInitMessage),
+	)
+	instance.Status.Conditions.Init(&cl)
+	instance.Status.ObservedGeneration = instance.Generation
+
+	defer func() {
+		// Restore condition timestamps if they haven't changed
+		condition.RestoreLastTransitionTimes(&instance.Status.Conditions, savedConditions)
+
+		if instance.Status.Conditions.IsUnknown(condition.ReadyCondition) {
+			instance.Status.Conditions.Set(instance.Status.Conditions.Mirror(condition.ReadyCondition))
+		}
+		if err := h.PatchInstance(ctx, instance); err != nil {
+			Log.Error(err, "Failed to patch instance")
+		}
+	}()
+
+	// Handle deletion
+	if !instance.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, instance, h)
+	}
+
+	// Add finalizer if not being deleted
+	if !controllerutil.ContainsFinalizer(instance, policyFinalizer) {
+		controllerutil.AddFinalizer(instance, policyFinalizer)
```

Comment:
this could simplified to:

```
if controllerutil.AddFinalizer(instance, policyFinalizer) {
  return ...
}
```

AddFinalizer accepts an Object and adds the provided finalizer if not present. It returns an indication of whether it updated the object's list of finalizers.